pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,223 @@
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer conn.Close()
+	infoResp, err := client.GetInfo(context.Background(), &watcherapi.InfoRequest{})
+	if err != nil {
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	resp := new(watcherapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[infoResp.Type]; ok {
+		var versions []string
+		for _, version := range infoResp.SupportedVersions {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(infoResp.Name, versions, socketPath); verifyErr != nil {
+			resp.PluginRegistered = false
+			resp.Error = fmt.Sprintf("Plugin registration failed with err: %v", verifyErr)
+		} else {
+			resp.PluginRegistered = true
+		}
+	} else {
+		resp.PluginRegistered = false
+		resp.Error = fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath)
+	}
+	glog.V(2).Infof("call NotifyRegistrationStatus, resp: %v", resp)
+	if _, err := client.NotifyRegistrationStatus(context.Background(), resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", infoResp.Type, infoResp.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", infoResp.Type, infoResp.Name, socketPath)
+	return nil
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+	w.watcher = watcher
+	startStatusCh := make(chan interface{})
+	go func(watcher *fsnotify.Watcher) {
+		close(startStatusCh)
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					w.registerPlugin(event.Name)
+				}
+				continue
+			case err := <-watcher.Errors:
+				//TODO: Handle errors by taking corrective measures
+				if err != nil {
+					glog.Errorf("Watcher received error: %v", err)
+				}
+				continue
+
+			case <-w.stopCh:
+				watcher.Close()
+				close(w.stopCh)
```

Comment:
As mentioned earlier use a stopChan + a waitgroup
Close the stopChan in the `Stop` call and then call `wg.Wait`

Reply:
What is issue with current approach?

Reply:
It's not maintainable, in the future if you add multiple channels you end up with the need to send as many messages as there are goroutine in the stopChan.
The golang idomatic way would be to close the stopChan and use a waitgroup that allows the code to be maintained and evolve correctly.

Reply:
Why would multiple stop channels be added? Can you please elaborate.

Reply:
Sorry correcting, "In the future if you add multiple **goroutines**".

Reply:
1. I am unable to think of a use case where one would ever need more than one goroutine for main watcher `for` loop. 
2. using waitgroup will make `Start` blocking and then kubelet will have to use goroutine to start it and also kubelet will have to use channels to get back any errors at watcher start, which is increased complexity compared to current approach.

Because of above points, would prefer to stick with current approach. Thanks!!

Reply:
> 1. I am unable to think of a use case where one would ever need more than one goroutine for main watcher for loop.

This review has already suggested to run a goroutine when registering an endpoint and I can see multiple ways to implement (recursive) sub-directories that involve goroutines

> 2. using waitgroup will make Start blocking and then kubelet will have to use goroutine to start it and also kubelet will have to use channels to get back any errors at watcher start, which is increased complexity compared to current approach.

No it won't:
```golang
func Start() {
    // init code here
    wg.Add(1)
    go func() {
        defer wg.Done()
        // fsnotify code here
   }()
}

func Stop() {
    close(stopChan)
    wg.Wait() // TODO: Timeout on wait
}
```